server/handler/uast_integration_test.go

```diff
@@ -2,19 +2,157 @@ package handler_test
 
 import (
 	"encoding/json"
+	"fmt"
 	"net/http"
 	"net/http/httptest"
 	"strings"
 	"testing"
 
+	"github.com/go-chi/chi"
 	"github.com/pressly/lg"
 	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
+	"gopkg.in/src-d/enry.v1"
 
 	"github.com/src-d/gitbase-web/server/handler"
 	"github.com/src-d/gitbase-web/server/serializer"
+	"github.com/src-d/gitbase-web/server/service"
 )
 
+type UASTGetLanguagesSuite struct {
+	suite.Suite
+	handler http.Handler
+}
+
+func TestUASTGetLanguagesSuite(t *testing.T) {
+	if !isIntegration() {
+		t.Skip("use the env var GITBASEPG_INTEGRATION_TESTS=true to run this test")
+	}
+
+	q := new(UASTGetLanguagesSuite)
+	r := chi.NewRouter()
+	r.Use(lg.RequestLogger(logrus.New()))
+	r.Post("/detect-lang", handler.APIHandlerFunc(handler.DetectLanguage()))
+	r.Get("/get-languages", handler.APIHandlerFunc(handler.GetLanguages(bblfshServerURL())))
+
+	q.handler = r
+
+	suite.Run(t, q)
+}
+
+func UnmarshalGetLanguagesResponse(b []byte) []service.Language {
+	var resBody struct {
+		Data []service.Language `json:"data"`
+	}
+	json.Unmarshal(b, &resBody)
+	return resBody.Data
+}
+
+func (suite *UASTGetLanguagesSuite) TestSameEnryLanguage() {
+	req, _ := http.NewRequest("GET", "/get-languages", strings.NewReader(""))
+
+	res := httptest.NewRecorder()
+	suite.handler.ServeHTTP(res, req)
+
+	suite.Require().Equal(http.StatusOK, res.Code, res.Body.String())
+
+	escapeForJSON := func(s string) string {
+		return strings.Replace(strings.Replace(s, "\"", "\\\"", -1),
+			"\n", "\\n", -1)
+	}
+
+	for _, lang := range UnmarshalGetLanguagesResponse(res.Body.Bytes()) {
+		langName := lang.Name
+		suite.T().Run(langName, func(t *testing.T) {
+			require := require.New(t)
+
+			content, filename := suite.getContentAndFilename(langName)
+			jsonRequest := fmt.Sprintf(`{ "content": "%s", "filename": "%s" }`,
+				escapeForJSON(content), filename)
```

Comment:
Instead of this manual escape you could do
```go
			j := struct {
				Filename string
				Content  string
			}{
				Filename: filename,
				Content:  content,
			}

			b, _ := json.Marshal(j)
			req, _ := http.NewRequest("POST", "/detect-lang", bytes.NewReader(b))
```

or export `detectLangRequest` type and
```go
			j := handler.DetectLangRequest{
				Filename: filename,
				Content:  content,
			}

			b, _ := json.Marshal(j)
			req, _ := http.NewRequest("POST", "/detect-lang", bytes.NewReader(b))
```

Reply:
thanks for the tip! Done [here](https://github.com/src-d/gitbase-web/pull/412/commits/b1a304cf5b4548b891612f47031ad8748f825ad9).